### PR TITLE
Provide native Java 11 HTTP client versions of `FormValidation#URLCheck` methods

### PR DESCRIPTION
**Pull request by basil**
Building on #7398, provide native Java 11 HTTP client versions of `FormValidation#URLCheck` methods, deprecating the old `HttpURLConnection` methods in favor of these and adding test coverage for the new functionality (the old functionality had no test coverage).

### Testing done

Ran `mvn clean verify -Dtest=hudson.util.FormValidationTest` locally. As a bonus, I converted `hudson.plugins.git.browser.AssemblaWebDoCheckURLTest#testInitialChecksOnRepoUrl` to this new API and verified that doing so chased away the weird Java 19 problem I was seeing in https://github.com/jenkinsci/bom/pull/1637#issuecomment-1345301126.

### Proposed changelog entries

Provide native Java 11 HTTP client versions of `FormValidation#URLCheck` methods

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7508"></a>

### CODE REVIEW

File: core/src/main/java/hudson/ProxyConfiguration.java
Comment (basil) [inline core/src/main/java/hudson/ProxyConfiguration.java:370]: Unrelated change while I was here: while testing with URLs like https://app.assembla.com/login that have a redirect, I noticed that the old `HttpURLConnection` API automatically followed redirects but the new Java 11 HTTP client API was not. To ease migration I figured it would be best to follow redirects by default. I could not think of a case where you would not want to follow them, and it matches the behavior of the old API. Of course anyone who wants to opt out can always do so by creating a custom builder without this option (something that was not possible with the old API).

File: core/src/main/java/hudson/util/FormValidation.java
Comment (basil) [inline core/src/main/java/hudson/util/FormValidation.java:555]: Minor unrelated change while I was here: append the stack trace for debuggability. We love stack traces.

**Review by timja — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Review by NotMyFault — APPROVED** [review]

**Pull request merged**
